yjit/src/compilation_log.rs

```diff
@@ -0,0 +1,242 @@
+use crate::core::BlockId;
+use crate::cruby::*;
+use crate::options::*;
+use crate::yjit::yjit_enabled_p;
+use crate::codegen::get_method_name;
+
+use std::fmt::{Display, Formatter};
+use std::os::raw::c_long;
+
+type Timestamp = f64;
+
+#[derive(Copy, Clone, Debug)]
+pub struct CompilationLogEntry {
+    /// The time when the block was compiled.
+    pub timestamp: Timestamp,
+
+    /// The compilation event payload.
+    pub payload: CompilationLogPayload,
+}
+
+#[derive(Copy, Clone, Debug)]
+pub enum CompilationLogPayload {
+    ISeq(BlockId),
+    CFunc(Option<VALUE>, ID)
+}
```

Comment:
Seems potentially risky to store BlockIds and `VALUE` in the compilation log. Some of these things can get GC'd and potentially move... Imo it would be safer to convert compilation log entries to strings as soon as they are added.

WDYT @k0kubun ?

Reply:
Yeah, the ISEQ of `BlockId` and `VALUE` need to be marked/updated on GC. It's possible but you often miss edge cases whenever you store `BlockId`.

Conversion to strings may bring allocation overhead. I think it's probably fine since we would run it only in the few servers that run `--yjit-stats`. But if we're interested in speeding it up, we could accept only limited characters for each entry and reuse the same string buffers, truncating entries if too long.

Reply:
If `BlockId` can be GC'd, that'll be a problem. I thought `CFunc` was only storing references to the core classes we substitute. But, I'll need to double-check that's accurate.

Eagerly resolving the string would avoid any of that headache. We'd just have to accept it'll consume more memory and there may be wasted resolutions if an entry is overwritten before retrieved from the circular buffer. If that's not a concern, I'll make the change.

Reply:
> If `BlockId` can be GC'd

`BlockId` is a pair of an ISEQ pointer and a u16 integer. The ISEQ address could be freed or moved.

Reply:
imo the simplest and safest option is to eagerly convert to a `String`. Since we only store 1024 entries by default, memory usage should not be a concern. We could be talking about ~128KiB when the compilation log is enabled.

Reply:
I was concerned not about the size but about the speed. With a String, I assume updating the String may require heap allocation every time and be somewhat slow. But yeah, it's probably fine too.

Reply:
I've updated the code to eagerly construct the `String` instances. However, I couldn't find a way to use a fixed-length buffer without resorting to `MaybeUninit`. Since memory optimization doesn't seem to be required here, I've moved it over to a `Vec` to keep the code simple.